// @angular-extensions/elements/fesm2015/angular-extensions-elements-testing.js
import * as i0 from '@angular/core';
import { Directive, NgModule } from '@angular/core';

class LazyElementTestingDirective {
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementTestingDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementTestingDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
LazyElementTestingDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.1.0", type: LazyElementTestingDirective, selector: "[axLazyElement]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementTestingDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[axLazyElement]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }]; } });

class LazyElementDynamicTestingDirective {
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementDynamicTestingDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementDynamicTestingDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
LazyElementDynamicTestingDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.1.0", type: LazyElementDynamicTestingDirective, selector: "[axLazyElementDynamic]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementDynamicTestingDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[axLazyElementDynamic]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }]; } });

class LazyElementsTestingModule {
}
LazyElementsTestingModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementsTestingModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LazyElementsTestingModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementsTestingModule, declarations: [LazyElementTestingDirective,
        LazyElementDynamicTestingDirective], exports: [LazyElementTestingDirective, LazyElementDynamicTestingDirective] });
LazyElementsTestingModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementsTestingModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementsTestingModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [
                        LazyElementTestingDirective,
                        LazyElementDynamicTestingDirective,
                    ],
                    exports: [LazyElementTestingDirective, LazyElementDynamicTestingDirective],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { LazyElementDynamicTestingDirective, LazyElementTestingDirective, LazyElementsTestingModule };
