// @angular-extensions/elements/esm2015/testing/lazy-element-dynamic-testing.directive.js
import { Directive, } from '@angular/core';
import * as i0 from "@angular/core";
export class LazyElementDynamicTestingDirective {
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementDynamicTestingDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementDynamicTestingDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
LazyElementDynamicTestingDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.1.0", type: LazyElementDynamicTestingDirective, selector: "[axLazyElementDynamic]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementDynamicTestingDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[axLazyElementDynamic]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }]; } });
