// @angular-extensions/elements/esm2015/testing/lazy-element-testing.directive.js
import { Directive, } from '@angular/core';
import * as i0 from "@angular/core";
export class LazyElementTestingDirective {
    constructor(vcr, template) {
        this.vcr = vcr;
        this.template = template;
    }
    ngOnInit() {
        this.vcr.createEmbeddedView(this.template);
    }
}
LazyElementTestingDirective.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementTestingDirective, deps: [{ token: i0.ViewContainerRef }, { token: i0.TemplateRef }], target: i0.ɵɵFactoryTarget.Directive });
LazyElementTestingDirective.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "12.1.0", type: LazyElementTestingDirective, selector: "[axLazyElement]", ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "12.1.0", ngImport: i0, type: LazyElementTestingDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[axLazyElement]',
                }]
        }], ctorParameters: function () { return [{ type: i0.ViewContainerRef }, { type: i0.TemplateRef }]; } });
